crates/agent/src/lib.rs

```diff
@@ -86,60 +85,50 @@ async fn listen_for_tasks(
     // This will then bubble through as a None return to PgListener:try_recv(),
     // which we'll then attempt to reconnect, and trigger the task handlers
     // to look and see if there's any work waiting for them.
-    let cloned_listener = listener.clone();
-    let handle = tokio::spawn(async move {
-        loop {
-            tokio::time::sleep(Duration::from_secs(30)).await;
-            cloned_listener
-                .lock()
-                .await
-                .listen("hacky_keepalive")
-                .await
-                .unwrap();
-
-            cloned_listener
-                .lock()
-                .await
-                .unlisten("hacky_keepalive")
-                .await
-                .unwrap();
-        }
-    });
-
     loop {
-        // try_recv returns None when the channel disconnects,
-        // which we want to have explicit handling for
-        let maybe_item = listener
-            .lock()
-            .await
-            .try_recv()
-            .await
-            .map_err(|e| anyhow::Error::from(e))?;
-
-        if let Some(item) = maybe_item {
-            let notification: AgentNotification = serde_json::from_str(item.payload())
-                .context("deserializing agent task notification")?;
-
-            tracing::debug!(
-                table = &notification.table,
-                "Message received to invoke handler"
-            );
-            task_tx
-                .send(HandlerInvocation {
-                    table_name: notification.table,
-                    is_poll: false,
-                })
-                .map_err(|e| anyhow::Error::from(e))?
-        } else {
-            tracing::warn!("LISTEN/NOTIFY stream from postgres lost, waking all handlers and attempting to reconnect");
-            table_names.iter().for_each(|table| {
-                task_tx
-                    .send(HandlerInvocation {
-                        table_name: table.clone(),
-                        is_poll: false,
-                    })
-                    .unwrap()
-            });
+        let should_poke_connection: Result<bool, anyhow::Error> = tokio::select! {
+            res = async {
+                tokio::time::sleep(Duration::from_secs(30)).await;
+                Ok(true)
+            } => {res},
+            res = async {
+                let maybe_item = listener.try_recv().await.map_err(|e| anyhow::Error::from(e))?;
```

Comment:
nit: I don't think you need map_err here and elsewhere. `?` should be doing that for you.

Reply:
You may want to use anyhow::Context to give more context for exactly where the error is coming from. Sometimes, if the error itself is terse, it can be hard to map it back to the code path it took to be bubbled up and I find `.context()` pretty invaluable there.

Reply:
I can honestly never tell when I need this map_error and when I don't. You're right, I don't need it here after all... but it certainly fixed a compiler error at some point /shrug